File: mobile/lib/services/receipt_outbox.dart
// Mobile receipt outbox.
//
// Why
// ---
// Receipts captured in low-signal places (basements, parking lots, planes)
// must never be lost. When /api/receipts/save fails, we queue the payload
// in shared_preferences and replay on next app launch / online window.
//
// Storage
// -------
// Single shared_preferences key 'getguac.outbox.v1' holding a JSON array.
// No Hive, no sqflite, no new packages — kept deliberately small so the
// mobile bundle stays lean (mobile-should-be-lightweight constraint).
//
// Each entry mirrors the web outbox shape:
//   { id, parsed, receipt_link, extra_page_urls, business_purchase,
//     validation_comment, user_category,
//     idempotency_key, attempts, last_error, queued_at }
//
// Flow
// ----
//   trySave(payload, sb)
//     1. POST /api/receipts/save with 30s timeout + Idempotency-Key.
//     2. 2xx → return SaveResult.success(receiptId, merged).
//     3. 4xx → throw (user fix needed, retry won't help).
//     4. Network / 5xx → enqueue + return SaveResult.queued(key).
//
//   flush(sb)
//     For each entry, POST it. 2xx / 4xx → remove. 5xx / network → keep
//     and bump attempts. Entries older than 7 days or with ≥10 attempts
//     are dropped with a warning.
//
// Auth: every POST forwards the user's Supabase access token as a Bearer
// header so the server-side /api/receipts/save can scope RLS correctly.

import 'dart:async';
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import 'package:supabase_flutter/supabase_flutter.dart';
import 'package:uuid/uuid.dart';

const String _kApiBase = 'https://getguac.app';
const String _kStorageKey = 'getguac.outbox.v1';
const int _kMaxAttempts = 10;
const Duration _kMaxAge = Duration(days: 7);
const Duration _kRequestTimeout = Duration(seconds: 30);

class SaveResult {
  final String? receiptId;
  final bool merged;
  final bool queued;
  final String? idempotencyKey;
  final String? error;

  const SaveResult._({
    this.receiptId,
    this.merged = false,
    this.queued = false,
    this.idempotencyKey,
    this.error,
  });

  factory SaveResult.success(String id, bool merged) =>
      SaveResult._(receiptId: id, merged: merged);
  factory SaveResult.queued(String key) =>
      SaveResult._(queued: true, idempotencyKey: key);
  factory SaveResult.failure(String err) => SaveResult._(error: err);
}

class ReceiptOutbox {
  static const _uuid = Uuid();

  /// Attempt a save. Never throws on network errors — queues instead.
  /// 4xx (user errors) still throw via the returned SaveResult.error.
  static Future<SaveResult> trySave(Map<String, dynamic> payload) async {
    final sb = Supabase.instance.client;
    final session = sb.auth.currentSession;
    if (session == null) {
      return SaveResult.failure('Not signed in');
    }
    final idemKey = (payload['idempotency_key'] as String?) ?? _uuid.v4();

    try {
      final res = await _post(idemKey, payload, session.accessToken);
      if (res.statusCode >= 200 && res.statusCode < 300) {
        final body = jsonDecode(res.body) as Map<String, dynamic>;
        return SaveResult.success(
          body['receipt_id'] as String,
          body['merged'] == true,
        );
      }
      if (res.statusCode >= 400 && res.statusCode < 500) {
        // 4xx — user fix needed. Don't queue; surface to caller.
        return SaveResult.failure(_decodeError(res));
      }
      // 5xx — queue + return
      await _enqueue(idemKey, payload, 'server ${res.statusCode}');
      return SaveResult.queued(idemKey);
    } catch (e) {
      // Network / timeout — queue + return.
      await _enqueue(idemKey, payload, e.toString());
      return SaveResult.queued(idemKey);
    }
  }

  /// Sweep the queue. Safe to call any time; never throws.
  /// Returns counts for diagnostics. Run from main() on app start +
  /// after each new save attempt.
  static Future<({int sent, int failed, int dropped})> flush() async {
    final sb = Supabase.instance.client;
    final session = sb.auth.currentSession;
    if (session == null) return (sent: 0, failed: 0, dropped: 0);

    final prefs = await SharedPreferences.getInstance();
    final list = _readList(prefs);
    if (list.isEmpty) return (sent: 0, failed: 0, dropped: 0);

    final now = DateTime.now().millisecondsSinceEpoch;
    final kept = <Map<String, dynamic>>[];
    int sent = 0, failed = 0, dropped = 0;

    for (final entry in list) {
      final queuedAt = (entry['queued_at'] as int?) ?? now;
      final attempts = (entry['attempts'] as int?) ?? 0;
      if (now - queuedAt > _kMaxAge.inMilliseconds || attempts >= _kMaxAttempts) {
        if (kDebugMode) {
          debugPrint('[outbox] dropping (TTL or attempts): ${entry['idempotency_key']}');
        }
        dropped++;
        continue;
      }

      final idemKey = entry['idempotency_key'] as String;
      try {
        final res = await _post(idemKey, entry, session.accessToken);
        if (res.statusCode >= 200 && res.statusCode < 300) {
          sent++;
          continue; // remove from queue
        }
        if (res.statusCode >= 400 && res.statusCode < 500) {
          if (kDebugMode) {
            debugPrint('[outbox] dropping 4xx ($idemKey): ${res.statusCode}');
          }
          dropped++;
          continue;
        }
        kept.add({
          ...entry,
          'attempts': attempts + 1,
          'last_error': 'server ${res.statusCode}',
        });
        failed++;
      } catch (e) {
        kept.add({
          ...entry,
          'attempts': attempts + 1,
          'last_error': e.toString(),
        });
        failed++;
      }
    }

    await _writeList(prefs, kept);
    return (sent: sent, failed: failed, dropped: dropped);
  }

  /// Read-only count for UI badges.
  static Future<int> size() async {
    final prefs = await SharedPreferences.getInstance();
    return _readList(prefs).length;
  }

  // ── Internals ─────────────────────────────────────────────────────

  static Future<http.Response> _post(
    String idemKey,
    Map<String, dynamic> entry,
    String accessToken,
  ) {
    final uri = Uri.parse('$_kApiBase/api/receipts/save');
    final body = jsonEncode({
      'parsed':             entry['parsed'],
      'receipt_link':       entry['receipt_link'],
      'extra_page_urls':    entry['extra_page_urls'],
      'business_purchase':  entry['business_purchase'],
      'validation_comment': entry['validation_comment'],
      'user_category':      entry['user_category'],
    });
    return http
        .post(
          uri,
          headers: {
            'Content-Type': 'application/json',
            'Authorization': 'Bearer $accessToken',
            'Idempotency-Key': idemKey,
          },
          body: body,
        )
        .timeout(_kRequestTimeout);
  }

  static String _decodeError(http.Response res) {
    try {
      final body = jsonDecode(res.body);
      if (body is Map && body['error'] is String) return body['error'] as String;
    } catch (_) {}
    return 'Save failed (${res.statusCode})';
  }

  static Future<void> _enqueue(
    String idemKey,
    Map<String, dynamic> payload,
    String lastError,
  ) async {
    final prefs = await SharedPreferences.getInstance();
    final list = _readList(prefs);
    final entry = <String, dynamic>{
      'idempotency_key':    idemKey,
      'parsed':             payload['parsed'],
      'receipt_link':       payload['receipt_link'],
      'extra_page_urls':    payload['extra_page_urls'],
      'business_purchase':  payload['business_purchase'],
      'validation_comment': payload['validation_comment'],
      'user_category':      payload['user_category'],
      'attempts':           1,
      'last_error':         lastError,
      'queued_at':          DateTime.now().millisecondsSinceEpoch,
    };
    // De-dup by idempotency key — replays must not double-enqueue.
    final filtered = list.where((e) => e['idempotency_key'] != idemKey).toList();
    filtered.add(entry);
    await _writeList(prefs, filtered);
  }

  static List<Map<String, dynamic>> _readList(SharedPreferences prefs) {
    final raw = prefs.getString(_kStorageKey);
    if (raw == null || raw.isEmpty) return const [];
    try {
      final decoded = jsonDecode(raw);
      if (decoded is List) {
        return decoded.whereType<Map<String, dynamic>>().toList();
      }
    } catch (e) {
      if (kDebugMode) debugPrint('[outbox] read failed: $e');
    }
    return const [];
  }

  static Future<void> _writeList(
    SharedPreferences prefs,
    List<Map<String, dynamic>> list,
  ) async {
    try {
      await prefs.setString(_kStorageKey, jsonEncode(list));
    } catch (e) {
      if (kDebugMode) debugPrint('[outbox] write failed: $e');
    }
  }
}

File: web/src/lib/receipt-outbox.js
// Web outbox for /api/receipts/save.
//
// Why
// ---
// User taps "Save" → network is flaky / offline / VPN switching. Without
// an outbox, the save just fails and the user has to remember to retry.
// With one, the save attempt either lands or queues, the UI never hangs,
// and a flush sweeps queued saves whenever the app reloads / regains net.
//
// Storage
// -------
// localStorage. Key: `getguac.outbox.v1`. Value: JSON array of entries.
// Mobile uses the same shape in shared_preferences so a future shared
// debug dump format works across platforms.
//
// Each entry:
//   { id, parsed, receipt_link, business_purchase, user_category,
//     validation_comment, extra_page_urls,
//     idempotency_key, attempts, last_error, queued_at }
//
// Flow
// ----
//   saveReceiptViaOutbox(opts)
//     1. Try POST /api/receipts/save with 30s timeout + Idempotency-Key.
//     2. On 2xx → return result. (Nothing queued; nothing to clean up.)
//     3. On network error / timeout / 5xx → enqueue, return { queued: true }.
//     4. On 4xx → throw (user error, retry won't help).
//
//   flushOutbox()
//     For each queued entry, replay it. On success → splice it out.
//     On 4xx → also splice out (it'll never succeed) and log.
//     On network/5xx → leave it, bump attempts.
//
// TTL
// ---
// Entries older than 7 days OR attempts >= 10 are auto-dropped on flush
// with a console warning. Keeps the queue from growing without bound.
//
// No new dependencies. No service worker (kept simple — flushes on load
// + post-save explicitly). No background sync (predictable behavior;
// service worker can come later).

const STORAGE_KEY = 'getguac.outbox.v1'
const MAX_ATTEMPTS = 10
const MAX_AGE_MS = 7 * 24 * 60 * 60 * 1000
const REQUEST_TIMEOUT_MS = 30_000

function readQueue() {
  if (typeof window === 'undefined') return []
  try {
    const raw = window.localStorage.getItem(STORAGE_KEY)
    if (!raw) return []
    const parsed = JSON.parse(raw)
    return Array.isArray(parsed) ? parsed : []
  } catch {
    return []
  }
}

function writeQueue(entries) {
  if (typeof window === 'undefined') return
  try {
    window.localStorage.setItem(STORAGE_KEY, JSON.stringify(entries))
  } catch (e) {
    // Quota exceeded — drop oldest to fit. Better than silently losing the
    // current save.
    console.warn('[outbox] localStorage write failed, trimming:', e.message)
    try {
      const trimmed = entries.slice(-20)
      window.localStorage.setItem(STORAGE_KEY, JSON.stringify(trimmed))
    } catch {}
  }
}

function newIdempotencyKey() {
  // crypto.randomUUID() returns 36 chars — well within the route's regex bounds.
  if (typeof crypto !== 'undefined' && crypto.randomUUID) return crypto.randomUUID()
  // Fallback: timestamp + random — also matches /^[A-Za-z0-9_-]{16,200}$/
  return `out-${Date.now()}-${Math.random().toString(36).slice(2, 12)}`
}

/**
 * Try to save. On network failure, enqueue + resolve with { queued: true }.
 * Never throws on network errors — the UI shouldn't have to handle them.
 * 4xx still throws (user fix needed, retry won't help).
 */
export async function saveReceiptViaOutbox(payload) {
  const idemKey = payload.idempotency_key || newIdempotencyKey()
  try {
    const res = await fetchWithTimeout('/api/receipts/save', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Idempotency-Key': idemKey,
      },
      body: JSON.stringify({
        parsed:             payload.parsed,
        receipt_link:       payload.receipt_link,
        extra_page_urls:    payload.extra_page_urls,
        business_purchase:  payload.business_purchase,
        validation_comment: payload.validation_comment,
        user_category:      payload.user_category,
      }),
    }, REQUEST_TIMEOUT_MS)

    if (res.status >= 200 && res.status < 300) {
      const json = await res.json()
      return { ...json, queued: false }
    }

    if (res.status >= 400 && res.status < 500) {
      const json = await res.json().catch(() => ({}))
      const err = new Error(json.error || `Save failed (${res.status})`)
      err.status = res.status
      throw err
    }

    // 5xx — network-class, queue + return
    throw new Error(`Server ${res.status}`)
  } catch (e) {
    // 4xx already threw above; everything else (network, timeout, 5xx) → queue.
    if (e.status && e.status >= 400 && e.status < 500) throw e
    enqueue({
      id: idemKey,
      parsed:             payload.parsed,
      receipt_link:       payload.receipt_link,
      extra_page_urls:    payload.extra_page_urls,
      business_purchase:  payload.business_purchase,
      validation_comment: payload.validation_comment,
      user_category:      payload.user_category,
      idempotency_key:    idemKey,
      attempts:           1,
      last_error:         e.message || 'network',
      queued_at:          Date.now(),
    })
    return { queued: true, idempotency_key: idemKey }
  }
}

function enqueue(entry) {
  const q = readQueue()
  // De-dup by idempotency_key — replays shouldn't create twin entries.
  const filtered = q.filter(e => e.idempotency_key !== entry.idempotency_key)
  filtered.push(entry)
  writeQueue(filtered)
}

async function fetchWithTimeout(url, init, ms) {
  // AbortSignal.timeout is widely available in modern browsers; fall back
  // to a manual controller for older ones.
  if (typeof AbortSignal !== 'undefined' && typeof AbortSignal.timeout === 'function') {
    return fetch(url, { ...init, signal: AbortSignal.timeout(ms) })
  }
  const ctrl = new AbortController()
  const t = setTimeout(() => ctrl.abort(), ms)
  try {
    return await fetch(url, { ...init, signal: ctrl.signal })
  } finally {
    clearTimeout(t)
  }
}

/**
 * Sweep the queue once. Returns { sent, failed, dropped } counts.
 * Safe to call any time. Caller should NOT await this in critical UI paths.
 */
export async function flushOutbox() {
  if (typeof window === 'undefined') return { sent: 0, failed: 0, dropped: 0 }
  if (typeof navigator !== 'undefined' && navigator.onLine === false) {
    return { sent: 0, failed: 0, dropped: 0 }
  }

  let q = readQueue()
  if (q.length === 0) return { sent: 0, failed: 0, dropped: 0 }

  const now = Date.now()
  const kept = []
  let sent = 0, failed = 0, dropped = 0

  for (const entry of q) {
    // TTL or attempt cap → drop with a warning. The user has presumably
    // moved on; the receipt is also still in the camera roll for re-capture.
    const ageMs = now - (entry.queued_at || now)
    if (ageMs > MAX_AGE_MS || (entry.attempts || 0) >= MAX_ATTEMPTS) {
      console.warn('[outbox] dropping entry (TTL or max attempts):', entry.idempotency_key, entry.last_error)
      dropped++
      continue
    }

    try {
      const res = await fetchWithTimeout('/api/receipts/save', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Idempotency-Key': entry.idempotency_key,
        },
        body: JSON.stringify({
          parsed:             entry.parsed,
          receipt_link:       entry.receipt_link,
          extra_page_urls:    entry.extra_page_urls,
          business_purchase:  entry.business_purchase,
          validation_comment: entry.validation_comment,
          user_category:      entry.user_category,
        }),
      }, REQUEST_TIMEOUT_MS)

      if (res.status >= 200 && res.status < 300) {
        sent++
        continue // drop from queue
      }
      if (res.status >= 400 && res.status < 500) {
        // User fix needed (e.g. RLS denied, validation). Retrying will keep
        // failing — drop with a log so the entry doesn't sit forever.
        console.warn('[outbox] dropping 4xx entry:', entry.idempotency_key, res.status)
        dropped++
        continue
      }
      // 5xx — keep, bump attempts
      kept.push({ ...entry, attempts: (entry.attempts || 0) + 1, last_error: `server ${res.status}` })
      failed++
    } catch (e) {
      // Network / timeout — keep, bump attempts
      kept.push({ ...entry, attempts: (entry.attempts || 0) + 1, last_error: e.message || 'network' })
      failed++
    }
  }

  writeQueue(kept)
  return { sent, failed, dropped }
}

/** Read-only count for UI badges. */
export function getOutboxSize() {
  return readQueue().length
}
